File: examples/adapters/trajectory_generators/jerk_limited/src/jerk_limited_generator.cpp
// Copyright 2025 Universidad Politécnica de Madrid
// SPDX-License-Identifier: BSD-3-Clause

/**
 * @file jerk_limited_generator.cpp
 * @author Rafael Perez-Segui <r.psegui@upm.es>
 */

#include "adapters/trajectory_generators/jerk_limited_generator.hpp"

#include <yaml-cpp/yaml.h>

#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <utility>

#include "utils/example_config_utils.hpp"
#include "utils/utils.hpp"

namespace mpc_examples::adapters {

namespace {

constexpr double kMinHorizontalSpeedForYaw = 0.5;   // [m/s]
constexpr double kMaxYawRateRefRadPerSec   = 1.0;   // [rad/s]

double wrapToPi(double x) {
  while (x > M_PI) x -= 2.0 * M_PI;
  while (x < -M_PI) x += 2.0 * M_PI;
  return x;
}

}  // namespace

JerkLimitedGenerator::JerkLimitedGenerator(const Config& cfg) : cfg_(cfg) {
  if (cfg_.reach_threshold <= 0.0) {
    throw std::invalid_argument("JerkLimitedGenerator: reach_threshold must be > 0.");
  }
}

JerkLimitedGenerator::Config JerkLimitedGenerator::loadConfigFromYaml(const std::string& path) {
  const YAML::Node root = detail::loadYamlRoot(path);
  Config cfg;
  const auto read_opt = [&](const char* key, double& dst) {
    if (root[key]) {
      dst = detail::readDoubleRequired(root[key], key);
    }
  };
  read_opt("max_acceleration", cfg.max_acceleration);
  read_opt("max_jerk", cfg.max_jerk);
  read_opt("max_tracking_error", cfg.max_tracking_error);
  read_opt("reach_threshold", cfg.reach_threshold);
  return cfg;
}

void JerkLimitedGenerator::initialize(const std::vector<Eigen::Vector3d>& waypoints,
                                      const mav_model::State& initial_state,
                                      const ExampleConfig& example_cfg) {
  if (waypoints.empty()) {
    throw std::invalid_argument("JerkLimitedGenerator: waypoints must not be empty.");
  }
  waypoints_    = waypoints;
  wp_index_     = 0U;
  finished_     = false;
  path_facing_  = example_cfg.path_facing;
  has_prev_t_   = false;
  prev_t_       = 0.0;
  yaw_ref_hold_ = quaternionToEuler(initial_state.getOrientationVector()).z();

  if (example_cfg.max_speed <= 0.0) {
    throw std::invalid_argument(
        "JerkLimitedGenerator: ExampleConfig::max_speed must be > 0 (set in config_example.yaml).");
  }
  trajectory_generator_jerk_limited::TrajectoryParameters params;
  params.max_speed          = example_cfg.max_speed;
  params.max_acceleration   = cfg_.max_acceleration;
  params.max_jerk           = cfg_.max_jerk;
  params.max_tracking_error = cfg_.max_tracking_error;

  ctrl_ = std::make_unique<trajectory_generator_jerk_limited::WaypointTrajectoryController>(params);
  const Eigen::Vector3d p0 = initial_state.getPositionVector();
  ctrl_->reset(p0);

  last_sample_.position     = p0;
  last_sample_.velocity     = Eigen::Vector3d::Zero();
  last_sample_.acceleration = Eigen::Vector3d::Zero();
  last_sample_.yaw          = yaw_ref_hold_;
}

void JerkLimitedGenerator::update(double t, const mav_model::State& state) {
  if (!ctrl_) {
    throw std::runtime_error("JerkLimitedGenerator: initialize() must be called before update().");
  }

  const double dt = has_prev_t_ ? std::max(t - prev_t_, 0.0) : 0.0;
  prev_t_         = t;
  has_prev_t_     = true;

  const Eigen::Vector3d position = state.getPositionVector();
  const Eigen::Vector3d target   = waypoints_[wp_index_];

  // Advance generator only when dt > 0 (the very first update() has dt==0 and
  // simply snapshots the initial setpoint).
  if (dt > 0.0) {
    const auto set = ctrl_->update(dt, position, target);
    last_sample_.position     = set.position;
    last_sample_.velocity     = set.velocity;
    last_sample_.acceleration = set.acceleration;
  }

  // Path-facing yaw with slew-rate limit (same policy as the legacy
  // mpc_trajectory_controller example).
  double yaw_used = yaw_ref_hold_;
  if (path_facing_) {
    const double horiz_speed = last_sample_.velocity.head<2>().norm();
    if (horiz_speed > kMinHorizontalSpeedForYaw) {
      const double yaw_target = std::atan2(last_sample_.velocity.y(), last_sample_.velocity.x());
      double yaw_delta        = wrapToPi(yaw_target - yaw_ref_hold_);
      const double step       = kMaxYawRateRefRadPerSec * (dt > 0.0 ? dt : 0.0);
      yaw_delta               = std::clamp(yaw_delta, -step, step);
      yaw_ref_hold_           = yaw_ref_hold_ + yaw_delta;
      yaw_used                = yaw_ref_hold_;
    }
  } else {
    yaw_used = 0.0;
  }
  last_sample_.yaw = yaw_used;

  // Waypoint advance when the drone enters the acceptance radius.
  const double err = (position - target).norm();
  if (err < cfg_.reach_threshold) {
    if (wp_index_ + 1U < waypoints_.size()) {
      ++wp_index_;
    } else {
      finished_ = true;
    }
  }
}

framework::ReferenceSample JerkLimitedGenerator::evaluate(double /*t*/) const {
  return last_sample_;
}

}  // namespace mpc_examples::adapters

File: examples/adapters/python/mpc_examples_adapters/trajectory_generators/jerk_limited_generator.py
#!/usr/bin/env python3

# Copyright 2025 Universidad Politécnica de Madrid
# SPDX-License-Identifier: BSD-3-Clause

"""Jerk-limited trajectory adapter.

Python mirror of
``examples/adapters/trajectory_generators/jerk_limited/src/jerk_limited_generator.cpp``.
Wraps :class:`trajectory_generator_jerk_limited.WaypointTrajectoryController`:
the internal generator advances in Python-space from ``t`` samples, waypoints
are retargeted when the drone enters the acceptance radius, and yaw follows
the filtered velocity with a rate-limited path-facing policy.
"""

__authors__ = 'Rafael Perez-Segui'
__license__ = 'BSD-3-Clause'

import math
import os
from dataclasses import dataclass
from typing import List, Optional

import numpy as np
import yaml
from mavpy.model import State
from trajectory_generator_jerk_limited import (
    TrajectoryParameters,
    WaypointTrajectoryController,
)

from mpc_examples_framework import (
    ExampleConfig,
    ITrajectoryGenerator,
    ReferenceField,
    ReferenceSample,
    make_mask,
)


_MIN_HORIZONTAL_SPEED_FOR_YAW = 0.5   # [m/s]
_MAX_YAW_RATE_REF_RAD_PER_SEC = 1.0   # [rad/s]


@dataclass
class JerkLimitedConfig:
    """Parsed configuration for :class:`JerkLimitedGenerator`.

    The 3D speed bound comes from :attr:`ExampleConfig.max_speed` at
    ``initialize()`` time; it is not exposed here on purpose so the full
    simulation stack agrees on a single source of truth.
    """

    max_acceleration: float = 0.0
    max_jerk: float = 0.0
    max_tracking_error: float = 0.0
    reach_threshold: float = 0.2


def _quat_to_yaw(q: np.ndarray) -> float:
    w, x, y, z = float(q[0]), float(q[1]), float(q[2]), float(q[3])
    siny_cosp = 2.0 * (w * z + x * y)
    cosy_cosp = 1.0 - 2.0 * (y * y + z * z)
    return math.atan2(siny_cosp, cosy_cosp)


def _wrap_to_pi(x: float) -> float:
    while x > math.pi:
        x -= 2.0 * math.pi
    while x < -math.pi:
        x += 2.0 * math.pi
    return x


class JerkLimitedGenerator(ITrajectoryGenerator):
    """Jerk-limited smoother over a waypoint list.

    The generator produces position + velocity + acceleration via the
    Ruckig-style limiter in ``trajectory_generator_jerk_limited``. The first
    update has ``dt == 0`` and just snapshots the initial setpoint; subsequent
    updates advance the internal state with the real wall-clock ``dt``.
    """

    def __init__(self, cfg: JerkLimitedConfig) -> None:
        if cfg.reach_threshold <= 0.0:
            raise ValueError('JerkLimitedGenerator: reach_threshold must be > 0.')
        self._cfg = cfg
        self._ctrl: Optional[WaypointTrajectoryController] = None
        self._waypoints: List[np.ndarray] = []
        self._wp_index = 0
        self._finished = False
        self._path_facing = False
        self._has_prev_t = False
        self._prev_t = 0.0
        self._yaw_ref_hold = 0.0
        self._last_sample = ReferenceSample()
        self._name = 'JerkLimitedGenerator'

    @staticmethod
    def load_config_from_yaml(path: str) -> JerkLimitedConfig:
        """Load a :class:`JerkLimitedConfig` from YAML.

        All keys are optional; the 3D speed bound is taken from
        :attr:`ExampleConfig.max_speed` at init time.
        """
        if not os.path.isfile(path):
            raise ValueError(f'Config file not found at {os.path.abspath(path)}.')
        with open(path, 'r') as f:
            root = yaml.safe_load(f)
        if root is not None and not isinstance(root, dict):
            raise ValueError('jerk_limited config: root must be a mapping.')
        cfg = JerkLimitedConfig()
        if isinstance(root, dict):
            for key in ('max_acceleration', 'max_jerk',
                        'max_tracking_error', 'reach_threshold'):
                if key in root:
                    setattr(cfg, key, float(root[key]))
        return cfg

    def initialize(self,
                   waypoints: List[np.ndarray],
                   initial_state: State,
                   example_cfg: ExampleConfig) -> None:
        if not waypoints:
            raise ValueError('JerkLimitedGenerator: waypoints must not be empty.')
        self._waypoints = [np.asarray(w, dtype=float) for w in waypoints]
        self._wp_index = 0
        self._finished = False
        self._path_facing = bool(example_cfg.path_facing)
        self._has_prev_t = False
        self._prev_t = 0.0
        self._yaw_ref_hold = _quat_to_yaw(np.asarray(initial_state.orientation, dtype=float))

        if example_cfg.max_speed <= 0.0:
            raise ValueError(
                'JerkLimitedGenerator: ExampleConfig.max_speed must be > 0 '
                '(set in config_example.yaml).')
        params = TrajectoryParameters()
        params.max_speed = example_cfg.max_speed
        params.max_acceleration = self._cfg.max_acceleration
        params.max_jerk = self._cfg.max_jerk
        params.max_tracking_error = self._cfg.max_tracking_error

        self._ctrl = WaypointTrajectoryController(params)
        p0 = np.asarray(initial_state.position, dtype=float)
        self._ctrl.reset(p0)

        self._last_sample = ReferenceSample(
            position=p0.copy(),
            velocity=np.zeros(3, dtype=float),
            acceleration=np.zeros(3, dtype=float),
            yaw=self._yaw_ref_hold,
        )

    def update(self, t: float, state: State) -> None:
        if self._ctrl is None:
            raise RuntimeError(
                'JerkLimitedGenerator: initialize() must be called before update().')

        dt = max(t - self._prev_t, 0.0) if self._has_prev_t else 0.0
        self._prev_t = t
        self._has_prev_t = True

        position = np.asarray(state.position, dtype=float)
        target = self._waypoints[self._wp_index]

        if dt > 0.0:
            setpoints = self._ctrl.update(dt, position, target)
            self._last_sample.position = np.asarray(setpoints.position, dtype=float).copy()
            self._last_sample.velocity = np.asarray(setpoints.velocity, dtype=float).copy()
            self._last_sample.acceleration = np.asarray(
                setpoints.acceleration, dtype=float).copy()

        if self._path_facing:
            horiz_speed = float(np.linalg.norm(self._last_sample.velocity[:2]))
            if horiz_speed > _MIN_HORIZONTAL_SPEED_FOR_YAW:
                yaw_target = math.atan2(float(self._last_sample.velocity[1]),
                                        float(self._last_sample.velocity[0]))
                yaw_delta = _wrap_to_pi(yaw_target - self._yaw_ref_hold)
                step = _MAX_YAW_RATE_REF_RAD_PER_SEC * (dt if dt > 0.0 else 0.0)
                yaw_delta = max(-step, min(step, yaw_delta))
                self._yaw_ref_hold += yaw_delta
            yaw_used = self._yaw_ref_hold
        else:
            yaw_used = 0.0
        self._last_sample.yaw = yaw_used

        err = float(np.linalg.norm(position - target))
        if err < self._cfg.reach_threshold:
            if self._wp_index + 1 < len(self._waypoints):
                self._wp_index += 1
            else:
                self._finished = True

    def evaluate(self, t: float) -> ReferenceSample:
        del t
        return ReferenceSample(
            position=self._last_sample.position.copy(),
            velocity=self._last_sample.velocity.copy(),
            acceleration=self._last_sample.acceleration.copy(),
            yaw=self._last_sample.yaw,
        )

    def is_finished(self, t: float) -> bool:
        del t
        return self._finished

    def current_waypoint_index(self) -> int:
        return int(self._wp_index)

    def provided_reference_fields(self) -> ReferenceField:
        return make_mask([
            ReferenceField.POSITION,
            ReferenceField.VELOCITY,
            ReferenceField.ACCELERATION,
        ])

    def name(self) -> str:
        return self._name
